### PR TITLE
[EOS-20328] Tuning of mempool to avoid read mempool buffers

### PR DESCRIPTION
**Pull request by SeagateChaDeepak**
# Change Summary
## Problem Statement/Description
_[EOS-20328](https://jts.seagate.com/browse/EOS-20328):_
_Tuning the read mempool to reduce memory consumption._

## Solution Overview
1. Removed array of buffers and kept a single 16k buffer size
2. Plugged 16k to write buffer where it is using from read mempool.

### Case 1:
Ran ST's to check the impact of changes on system.
MP Job : http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/1073/console
Pre-Merge Job : http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/1087/console

### Case 2:
Parallelly uploaded 2 100 mb files to s3 server to check if there is any issue. 

Signed-off-by: Deepak Mahale <deepak.mahale@seagate.com>

### CODE REVIEW

File: s3config.yaml.sample [hunk 1]
```diff
@@ -98,9 +98,9 @@ S3_MOTR_CONFIG:                                     # Section for S3 Motr
    S3_MOTR_CASS_CLUSTER_EP: 127.0.0.1                 # Cassandra cluster end point
    S3_MOTR_CASS_KEYSPACE: "motr_index_keyspace"     # Cassandra keyspace
    S3_MOTR_CASS_MAX_COL_FAMILY_NUM: 1                 # Cassandra max column family number, default is 1
-   S3_UNIT_SIZES_FOR_MEMORY_POOL: [4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1048576] # Memory pool will be created for each of these unit_size with following properties
-   S3_MOTR_READ_POOL_INITIAL_BUFFER_COUNT: 10        # 10 blocks, the initial pool size = multiple of S3_MOTR_UNIT_SIZE
-   S3_MOTR_READ_POOL_EXPANDABLE_COUNT: 50            # 50 blocks, pool's expandable size, multiple of S3_MOTR_UNIT_SIZE
+   S3_UNIT_SIZES_FOR_MEMORY_POOL: [16384]              # Memory pool will be created for each of these unit_size with following properties
+   S3_MOTR_READ_POOL_INITIAL_BUFFER_COUNT: 3         # 10 blocks, the initial pool size = multiple of S3_MOTR_UNIT_SIZE
```
Comment (knrajnambiar76) [inline s3config.yaml.sample:102]: you can have it 10 itself

Comment (SeagateChaDeepak) [reply s3config.yaml.sample:102]: 10 and 50?

**Review by knrajnambiar76 — APPROVED** [review]

File: s3config-test.yaml [hunk 2]
```diff
@@ -98,9 +98,9 @@ S3_MOTR_CONFIG:                                     # Section for S3 Motr
    S3_MOTR_CASS_CLUSTER_EP: 10.10.1.3                 # Cassandra cluster end point
    S3_MOTR_CASS_KEYSPACE: "motr_index_keyspace"     # Cassandra keyspace
    S3_MOTR_CASS_MAX_COL_FAMILY_NUM: 1                 # Cassandra max column family number, default is 1
-   S3_UNIT_SIZES_FOR_MEMORY_POOL: [4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1048576, 2097152, 4194304]                          # Memory pool will be created for each of this unit_size with following properties
-   S3_MOTR_READ_POOL_INITIAL_BUFFER_COUNT: 10         # 10 blocks, the initial pool size, multiple of S3_MOTR_UNIT_SIZE
-   S3_MOTR_READ_POOL_EXPANDABLE_COUNT: 20             # 20 blocks, pool's expandable size, multiple of S3_MOTR_UNIT_SIZE
+   S3_UNIT_SIZES_FOR_MEMORY_POOL: [16384]              # Memory pool will be created for each of this unit_size with following properties
+   S3_MOTR_READ_POOL_INITIAL_BUFFER_COUNT: 10          # 10 blocks, the initial pool size, multiple of S3_MOTR_UNIT_SIZE
+   S3_MOTR_READ_POOL_EXPANDABLE_COUNT: 50             # 20 blocks, pool's expandable size, multiple of S3_MOTR_UNIT_SIZE
```
Comment (amitwac2608) [inline s3config-test.yaml:103]: I think comment needs to be fixed to have '50 blocks'

File: s3config-test.yaml [hunk 3]
```diff
@@ -98,9 +98,9 @@ S3_MOTR_CONFIG:                                     # Section for S3 Motr
    S3_MOTR_CASS_CLUSTER_EP: 10.10.1.3                 # Cassandra cluster end point
    S3_MOTR_CASS_KEYSPACE: "motr_index_keyspace"     # Cassandra keyspace
    S3_MOTR_CASS_MAX_COL_FAMILY_NUM: 1                 # Cassandra max column family number, default is 1
-   S3_UNIT_SIZES_FOR_MEMORY_POOL: [4096, 8192, 16384, 32768, 65536, 131072, 262144, 524288, 1048576, 2097152, 4194304]                          # Memory pool will be created for each of this unit_size with following properties
-   S3_MOTR_READ_POOL_INITIAL_BUFFER_COUNT: 10         # 10 blocks, the initial pool size, multiple of S3_MOTR_UNIT_SIZE
-   S3_MOTR_READ_POOL_EXPANDABLE_COUNT: 20             # 20 blocks, pool's expandable size, multiple of S3_MOTR_UNIT_SIZE
+   S3_UNIT_SIZES_FOR_MEMORY_POOL: [16384]              # Memory pool will be created for each of this unit_size with following properties
```
Comment (amitwac2608) [inline s3config-test.yaml:101]: Now we have single unit size, do we still need to show it as array ?

Also we can update comment to:
"Memory pool will be created for this size with following properties."

File: server/s3_motr_writer.cc [hunk 4]
```diff
@@ -699,10 +699,8 @@ void S3MotrWiter::set_up_motr_data_buffers(struct s3_motr_rw_op_context *rw_ctx,
   if (buf_idx < motr_buf_count) {
     // Allocate place_holder_for_last_unit only if its required.
     // Its required when writing last block of object.
+    unit_size_for_place_holder = 16384;
```
Comment (amitwac2608) [inline server/s3_motr_writer.cc:702]: I would suggest o declare a const for 16384, and use that const here.

File: s3config-test.yaml (hunk 3 above)
Comment (SeagateChaDeepak) [reply s3config-test.yaml:101]: Will address these changes in https://jts.seagate.com/browse/EOS-20630

File: server/s3_motr_writer.cc (hunk 4 above)
Comment (SeagateChaDeepak) [reply server/s3_motr_writer.cc:702]: Will address these changes in https://jts.seagate.com/browse/EOS-20630

File: s3config-test.yaml (hunk 2 above)
Comment (SeagateChaDeepak) [reply s3config-test.yaml:103]: Will Address these changes in https://jts.seagate.com/browse/EOS-20630

**Review by amitwac2608 — APPROVED** [review]

**Pull request merged**